src/impl/FormulaCell.ts

```diff
@@ -90,105 +90,299 @@ export class FormulaCell<T> implements Cell<T>, Recalculable {
     }
 }
 
-export function formula<S1, T>(
-    fn: (val1: S1) => T,
-    input1: Value<S1>
-): FormulaCell<T> | T
-export function formula<S1, S2, T>(
-    fn: (val1: S1, val2: S2) => T,
-    input1: Value<S1>,
-    input2: Value<S2>
-): FormulaCell<T> | T
-export function formula<S1, S2, S3, T>(
-    fn: (val1: S1, val2: S2, val3: S3) => T,
-    input1: Value<S1>,
-    input2: Value<S2>,
-    input3: Value<S3>
-): FormulaCell<T> | T
-export function formula<S1, S2, S3, S4, T>(
-    fn: (val1: S1, val2: S2, val3: S3, val4: S4) => T,
-    input1: Value<S1>,
-    input2: Value<S2>,
-    input3: Value<S3>,
-    input4: Value<S4>
-): FormulaCell<T> | T
-export function formula<S1, S2, S3, S4, S5, T>(
-    fn: (val1: S1, val2: S2, val3: S3, val4: S4, val5: S5) => T,
-    input1: Value<S1>,
-    input2: Value<S2>,
-    input3: Value<S3>,
-    input4: Value<S4>,
-    input5: Value<S5>
-): FormulaCell<T> | T
-export function formula<S1, S2, S3, S4, S5, S6, T>(
-    fn: (val1: S1, val2: S2, val3: S3, val4: S4, val5: S5, val6: S6) => T,
-    input1: Value<S1>,
-    input2: Value<S2>,
-    input3: Value<S3>,
-    input4: Value<S4>,
-    input5: Value<S5>,
-    input6: Value<S6>
-): FormulaCell<T> | T
-export function formula<S1, S2, S3, S4, S5, S6, S7, T>(
+export type UnwrapCell<T> = T extends Cell<infer U> ? U : T
+
+export function formula<S1 extends Value<unknown>, T, U1 = UnwrapCell<S1>>(
```

Comment:
Not quite get it, why there are `unknown` and `any`?

Reply:
Let's just look at the first overload:
```typescript
export function formula<S1 extends Value<unknown>, T, U1 = UnwrapCell<S1>>(
    fn: (val1: U1) => T,
    input1: S1
): S1 extends Cell<any> ? FormulaCell<T> : T
```
With `S1 extends Value<unknown>` we add a [constraint to the generic type variable](https://www.typescriptlang.org/docs/handbook/2/generics.html#generic-constraints) `S1`. This way `S1` cannot be anything - it is required to be of type `Value` of something currently unknown (`S1 extends Value<any>` would work the same way - I just find `unknown` more expressive). Instead of `unknown` I would prefer to write `S1 extends Value<U1>` to get `S1` as type variable of the `Value` container and `U1` as type variable of the value in the `Value` container. But that's not how typescript works (as far as I know). 
To get  the type of the value in the `Value` container we use the `UnwrapCell` helper to [`infer` it](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-8.html#type-inference-in-conditional-types) from `Cell`. 
This way we can use `S1` as type for `input1` which could be for example `Cell<number>` or just `number`. And we have `U1` which will always be `number` in this example. We use this as type for `val1`.

Finally we can use `S1` to get the correct `ReturnType` of `formula` function as [a conditional type](https://www.typescriptlang.org/docs/handbook/2/conditional-types.html). Let's still assume `S1` is of type `Value<number>`. If `S1` extends from any kind of `Cell` (in this example it would be `Cell<number>` but it actually could be `any` `Cell`), the `ReturnType` will be `FormulaCell<number>`. If `S1` is not of `Cell`, the `ReturnType` will be just `number`.

Now I see that the `ReturnType` could also be written more strict if you prefer: 
```typescript
S1 extends Cell<U1> ? FormulaCell<T> : T
```
But it won't be necessary because if we stick with the `number` example, `S1` can only be of type `Cell<number>` or `number`. So if `S1` extends of `any` kind of `Cell` (`Cell<any>`) the `ReturnType` will be a `FormulaCall`.

Reply:
This is awesome @pabra . Great job!